include/revng/EarlyFunctionAnalysis/ControlFlowGraph.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <concepts>
+#include <map>
+
+#include "revng/ADT/Concepts.h"
+#include "revng/ADT/GenericGraph.h"
+#include "revng/EarlyFunctionAnalysis/FunctionMetadata.h"
+#include "revng/Model/FunctionAttribute.h"
+
+namespace efa {
+
+namespace detail {
+using SuccessorContainer = SortedVector<UpcastablePointer<FunctionEdgeBase>>;
+}
+
+template<typename T>
+concept IsBasicBlock = requires {
+  { T::Start } -> convertible_to<MetaAddress>;
+  { T::End } -> convertible_to<MetaAddress>;
+  { T::Successors } -> convertible_to<detail::SuccessorContainer>;
+};
+
+struct ParsedSuccessor {
+  MetaAddress NextInstructionAddress;
+  MetaAddress OptionalCallAddress;
+};
+
+inline ParsedSuccessor parseSuccessor(const efa::FunctionEdgeBase &Edge,
+                                      const MetaAddress &FallthroughAddress,
+                                      const model::Binary &Binary) {
+  switch (Edge.Type) {
+  case FunctionEdgeType::DirectBranch:
+  case FunctionEdgeType::FakeFunctionCall:
+  case FunctionEdgeType::FakeFunctionReturn:
+  case FunctionEdgeType::Return:
+  case FunctionEdgeType::BrokenReturn:
+  case FunctionEdgeType::IndirectTailCall:
+  case FunctionEdgeType::LongJmp:
+  case FunctionEdgeType::Unreachable:
+    return ParsedSuccessor{ .NextInstructionAddress = Edge.Destination,
+                            .OptionalCallAddress = MetaAddress::invalid() };
+
+  case FunctionEdgeType::FunctionCall:
+  case FunctionEdgeType::IndirectCall:
+    if (auto *CE = llvm::cast<efa::CallEdge>(&Edge);
+        !hasAttribute(Binary, *CE, model::FunctionAttribute::NoReturn)) {
+      return ParsedSuccessor{ .NextInstructionAddress = FallthroughAddress,
+                              .OptionalCallAddress = Edge.Destination };
+    } else {
+      return ParsedSuccessor{ .NextInstructionAddress = MetaAddress::invalid(),
+                              .OptionalCallAddress = Edge.Destination };
+    }
+
+  case FunctionEdgeType::Killer:
+    return ParsedSuccessor{ .NextInstructionAddress = MetaAddress::invalid(),
+                            .OptionalCallAddress = MetaAddress::invalid() };
+
+  default:
+  case FunctionEdgeType::Invalid:
+  case FunctionEdgeType::Count:
+    revng_abort();
+    break;
+  }
+}
+
+// clang-format off
+template<IsGenericGraph GraphType,
+         IsBasicBlock BasicBlockType,
+         typename ...OtherTs,
+         template<typename...> typename Container>
+  requires std::is_constructible_v<typename GraphType::Node,
+                                   const MetaAddress &>
+std::pair<GraphType, std::map<MetaAddress, typename GraphType::Node *>>
+buildControlFlowGraph(const Container<BasicBlockType, OtherTs...> &BasicBlocks,
+                      const MetaAddress &EntryAddress,
+                      const model::Binary &Binary) {
```

Comment:
See if you can make this more accessible and add some doc.

Reply:
Always a good idea! :)

Reply:
Something like this? https://github.com/revng/revng/pull/179/commits/17c6a939a0f08d8e81eb151378dbadbb6ef878d9

Reply:
Yeah!